### ISSUE TITLE
Offset between cone positions in the ground truth topic and Unreal Engine

### ISSUE THREAD

**Issue by davidoort** [FS-Driverless/Formula-Student-Driverless-Simulator#144]
On training map (I haven't tested other maps), commit cd0ffe64d15cdc1f5e26f8a46cc8d490e5d81fef, the following can be observed.

![image](https://user-images.githubusercontent.com/21220927/87307230-7b32c300-c519-11ea-8b90-2ad711d3aec1.png)

**Comment by SijmenHuizenga** [comment]
Which commit is this?

**Comment by SijmenHuizenga** [comment]
Is it coincidence that the offset looks like to be the width of the vehicle?

**Comment by SijmenHuizenga** [comment]
@davidoort Do you have an idea how we can test this without the FSTD autonomous system? I would like to confirm this is a FSDS. In the picture you show it could also be an FSTD integration problem.

**Comment by davidoort** [comment]
@SijmenHuizenga the offset seems to only occur in the x direction (i.e. the direction the car is pointing to at the start). This indicates that if there is a problem, it could do with the initial x car position in unreal (or how that position is defined vs how we interpret it to be defined). In Referee.cpp we load this start_pos (which comes from the Refereeblueprint) which is subtracted from all the cone positions. I interpret that position as the CG of the car, but maybe it is another weird reference point on the vehicle.
![image](https://user-images.githubusercontent.com/21220927/87454141-dc36c580-c603-11ea-895d-9015b01b6e72.png)

>  Do you have an idea how we can test this without the FSTD autonomous system? I would like to confirm this is a FSDS. In the picture you show it could also be an FSTD integration problem.

I can't think of an easy way that doesn't require creating a visualization publisher and a map to car  frame broadcaster based on odometry.

It would be really helpful if for now you could clarify what this start pos actually is

**Comment by smnschfr** [comment]
An x-offset of 2 meters seems to match pretty well.
This was just obtained through trying out, so no idea where the error lies.

**Comment by SijmenHuizenga** [comment]
I found a way to test it easely:

1. Create a map in unreal engine with only 4 cones

2. Add the following debug lines to the wrapper:

```
diff --git a/ros/src/fsds_ros_bridge/src/airsim_ros_wrapper.cpp b/ros/src/fsds_ros_bridge/src/airsim_ros_wrapper.cpp
index 80dcd5c..5d764a2 100644
--- a/ros/src/fsds_ros_bridge/src/airsim_ros_wrapper.cpp
+++ b/ros/src/fsds_ros_bridge/src/airsim_ros_wrapper.cpp
@@ -83,6 +83,9 @@ void AirsimROSWrapper::publish_track() {
     // Get car initial position
     car_start_pos = state.car_start_location;

+^M
+    std::cout << "car" << " x: " << car_start_pos.x << " y: " << car_start_pos.y << std::endl;^M
+^M
     fs_msgs::Track track;
     for (const auto& cone : state.cones) {
         fs_msgs::Cone cone_object;
@@ -90,6 +93,7 @@ void AirsimROSWrapper::publish_track() {
         cone_object.location.x = cone.location.x != 0 ? (cone.location.x - car_start_pos.x)*0.01 : 0;
         // Negative sign follows ENU convention
         cone_object.location.y = cone.location.y != 0 ? -(cone.location.y - car_start_pos.y)*0.01 : 0;
+        std::cout << "cone" << " x: " << cone_object.location.x << " y: " << cone_object.location.y << " color: " << cone.color << std::endl;^M
         if (cone.color == CarApiBase::ConeColor::Yellow) {
             cone_object.color = fs_msgs::Cone::YELLOW;
         } else if (cone.color == CarApiBase::ConeColor::Blue) {
```

3. Use the measuring tool in unreal editor to check the distance between car and cones
4. profit

**Comment by davidoort** [comment]
According to recent tests, this issue does not seem to have been fixed in either map. Is there a plan to fix this before competition?
![image (20)](https://user-images.githubusercontent.com/21220927/89122150-5ea8fb80-d4c5-11ea-8f6e-28290a15ac93.png)
![image (19)](https://user-images.githubusercontent.com/21220927/89122152-5fda2880-d4c5-11ea-8d17-e6e9d7eb37c3.png)

**Comment by SijmenHuizenga** [comment]
When testing on the trainingmap with the following cones:

![image](https://user-images.githubusercontent.com/8008227/89122665-5d79cd80-d4c9-11ea-8a62-f02b292f60a2.png)


So the cones relative to the car should be at x 284

![ScreenShot00000](https://user-images.githubusercontent.com/8008227/89122683-a7fb4a00-d4c9-11ea-8920-acb196f25788.png)


However, they are at 4.8 (using the above print statements output the following)

```
car x: 4575.15 y: 8577.82
cone x: 4.82257 y: -2.06277 color: 0
cone x: 4.82257 y: 1.43723 color: 1
```


So yea you are right, the problem still exists.

**Comment by davidoort** [comment]
Thanks for creating this test setup! We can either find the root source of the issue in Unreal Engine or we can simply correct for this known offset (198.257 cm) in the ROS wrapper (which at least would make it ready by the competition)

**Comment by smnschfr** [comment]
Please let us know if you change anything here before the competion. Right now, we have implemented a correction for this offset in our AS which would need to be reversed once it's fixed in the signal.

**Comment by SijmenHuizenga** [comment]
Apparently the vehicle is spawned offset from the playerstart. 

The arrows show the playerstart component, the vehicle is untouched.
![image](https://user-images.githubusercontent.com/8008227/89357012-9aa7b080-d6bf-11ea-964c-2d3191073a9e.png)

The offset is probably caused by `void ASimModeBase::setupVehiclesAndCamera()`. This all is because airsim supports changing the spawn position from settings.jon.

**Comment by mateusz-lichota** [comment]
This was fixed in #274
